canvas/gradient.go

```diff
@@ -0,0 +1,176 @@
+package canvas
+
+import (
+	"errors"
+	"image"
+	"image/color"
+	"image/draw"
+	"math"
+
+	"fyne.io/fyne"
+)
+
+const (
+	// GradientDirectionVertical defines a vertically traveling gradient
+	GradientDirectionVertical GradientDirection = iota
+	// GradientDirectionHorizontal defines a horizontally traveling gradient
+	GradientDirectionHorizontal
+	// GradientDirectionCircular defines a gradient traveling radially from a center point outward
+	GradientDirectionCircular
+)
+
+// GradientDirection defines a starting or stopping location
+type GradientDirection int
+
+// Gradient describes a gradient fade between two colors
+type Gradient struct {
+	baseObject
+
+	Generator func(w, h int) image.Image
+
+	direction GradientDirection // The direction of the gradient.
+	start     color.Color       // The beginning RGBA color of the gradient
+	end       color.Color       // The end RGBA color of the gradient
+	center    fyne.Position     // The offest for generation of the gradient
+
+	gradientFnc func(x, y, w, h int, ox, oy float64) float64 // The function for calculating the gradient
+
+	img draw.Image // internal cache for pixel generator - may be superfluous
+}
+
+// Center returns the center of the gradient
+func (g *Gradient) Center() fyne.Position {
+	return g.center
+}
+
+// SetCenter sets the gradient center.
+// Should be set before generation. Only used by Circular
+func (g *Gradient) SetCenter(pos fyne.Position) {
+	g.center = pos
+}
+
+// calculatePixel uses the gradientFnc to caculate the pixel
+// at x, y as a gradient between start and end
+// using w and h to determine rate of graduation
+// returns a color.RGBA64
+func (g *Gradient) calculatePixel(w, h, x, y int) *color.RGBA64 {
+	ox, oy := float64(g.center.X), float64(g.center.Y)
+	d := g.gradientFnc(x, y, w, h, ox, oy)
+
+	// fetch RGBA values
+	aR, aG, aB, aA := g.start.RGBA()
+	bR, bG, bB, bA := g.end.RGBA()
+
+	// Get difference
+	dR := (float64(bR) - float64(aR))
+	dG := (float64(bG) - float64(aG))
+	dB := (float64(bB) - float64(aB))
+	dA := (float64(bA) - float64(aA))
+
+	// Apply graduation
+	pixel := &color.RGBA64{
+		R: uint16(float64(aR) + d*dR),
+		B: uint16(float64(aB) + d*dB),
+		G: uint16(float64(aG) + d*dG),
+		A: uint16(float64(aA) + d*dA),
+	}
+
+	// Prevent rendering dead space - transparent corners
+	if d > 1 {
+		pixel.A = uint16(0)
+	}
+
+	return pixel
+
+}
+
+// pixelGradient is used during construction
+type pixelGradient struct {
+	g   *Gradient
+	img draw.Image // image cache during generator run
+}
+
+// NewLinearGradient returns a new Image instance that dynamically
+// renders a gradient of type GradientDirection
+func NewLinearGradient(start color.Color, end color.Color, direction GradientDirection) (*Gradient, error) {
+	if start == nil || end == nil {
+		return nil, errors.New("start and end colors are required for a gradient")
```

Comment:
The model used by other CanvasObject is to set any value and to handle errors when rendering.
I don't know specifically whether CanvasObject constructors should return an error value - but the others have the same potential issue but handle it differently. Is consistency in this reasonable?